Feature/model too many joins

BradCr opened the pull request:
This is a:
- [ ] bug fix PR with no breaking changes
- [x] new functionality

## Link to Issue 
Closes [#394](https://github.com/dbt-labs/dbt-project-evaluator/issues/394)

## Description & motivation
Per the dbt [best practices docs](https://docs.getdbt.com/best-practices/how-we-structure/3-intermediate#intermediate-models), models should bring together a reasonable number (typically 4 to 6) of entities or concepts (staging models, or perhaps other intermediate models) that will be joined with another similarly purposed intermediate model to generate a mart. Having too many models in our mart increases the complexity. We can join two intermediate models that each house a piece of the complexity, giving us increased readability, flexibility, testing surface area, and insight into our components.

This new model, `fct_number_of_joins`, will identify models that join from seven (7) or more other models and should be refactored into intermediate models each taking some of the joins.

## Integration Test Screenshot
The equality test for the new `fct_number_of_joins` model passed in the integration tests.

![dbt test fct_number_of_joins](https://github.com/dbt-labs/dbt-project-evaluator/assets/151274228/31b78c56-94c4-4ec7-b1f2-6a1f182aa539)

Screenshot of the `fct_number_of_joins` model being built:

![dbt_fct_number_of_joins_build](https://github.com/dbt-labs/dbt-project-evaluator/assets/151274228/e348aedf-d06b-426b-9b8b-d72d79096f41)

I updated the documentation to explain the new model. For the screenshots I had two concerns; When I attached the screenshots they are linked to my repository (not sure if that gets automatically updated or if it's something that a dbt maintainer would have to deal with) and the screenshots are from the new Cloud Lineage diagram, so the look and feel is different from the existing screenshots.

## Checklist
- [x] I have verified that these changes work locally on the following warehouses (Note: it's okay if you do not have access to all warehouses, this helps us understand what has been covered)
    - [ ] BigQuery
    - [ ] Postgres
    - [ ] Redshift
    - [ ] Snowflake
    - [ ] Databricks
    - [x] DuckDB
    - [ ] Trino/Starburst
- [ ] I have updated the README.md (if applicable)
- [x] I have added tests & descriptions to my models (and macros if applicable)

BradCr commented:
The integration tests look like they're failing because, now that I've introduced this new model, some test models, and the associated seed, it is failing on some other equality tests. I assume I need to update the contents of those seed files to include what I've added.

Is that the correct approach?

b-per commented:
Yes, this is a bit annoying, but we need to update a couple of seed files when adding new models in the `integration_tests` folder. So, please, feel free to update the seeds for the tests that started failing.

This doesn't happen every day so we kept it like this for now, but we know that it is not fun to do 😄

graciegoheen commented:
Hey! How would you feel about naming the model `fct_too_many_joins` or something similar so that the model name describes the problem being flagged?

BradCr commented:
@b-per Thanks for clarifying, I'll align all the seeds so the equality tests no longer fail. I think I saw a couple discrepancies that weren't just about my new model (when I tested locally in VSCode) so I'll double-check for those as well.

@graciegoheen That's a good suggestion. Now that you mention it, I'm not sure why I didn't name the model that to begin with, seeing as I named both the branch and the PR "too many joins".

BradCr commented:
Renamed the model from `fct_number_of_joins` to `fct_too_many_joins`. Next I'll work on updating the seeds for the integration test.

BradCr commented:
@b-per @graciegoheen

I've updated the seeds where I saw differences, but two of the tests that are failing on the integration tests here on this PR, `equality_fct_undocumented_models` and `equality_fct_documentation_coverage`, both pass when I run them locally:

![image](https://github.com/dbt-labs/dbt-project-evaluator/assets/151274228/4b954be7-b096-4632-a26a-8c11986a799a)

I'm not sure how to get past this point.

b-per commented:
Thanks @BradCr , we will have a look at the CI side

b-per commented:
OK, so the tests are also failing for me locally, this is not just CI.

Can you do a `dbt build -f` (to full refresh the seeds as well) in your `integration_tests` folder and check if it all works for you locally?

I am thinking that in your screenshots, the tests are passing but it is because the data they use has not been refreshed from your latest changes in the seeds files.

BradCr commented:
Thanks for checking into it. I'l do the full rebuild and then run the tests again.

BradCr commented:
@b-per I ran `dbt build -f` in the integration_tests folder, and I still see the same thing. On my local those two equality tests are passing, but failing on the circleci check. I then even deleted the duckdb file and ran the full build again so it would recreate all the tables from scratch, and still those two tests are passing on my local.

When I run the compile SQL for those two tests on the duckdb database they pass, When I query the fct_ and test_fct_ tables individually, the values in the rows are equal.

However, I do see that there are some other tests failing on my local that I'll need to look at, so I will try to sort those out in the meatime:

```
23:56:14  Failure in test equality_fct_source_directories (seeds\structure\structure_seeds.yml)
23:56:14  Failure in test equality_fct_test_directories (seeds\structure\structure_seeds.yml)
23:56:14  Failure in test equality_fct_chained_views_dependencies (seeds\performance\performance_seeds.yml)
23:56:14  Failure in test equality_fct_model_directories (seeds\structure\structure_seeds.yml)
23:56:14  Failure in test equality_fct_model_fanout (seeds\dag\dag_seeds.yml)
23:56:14  Failure in test equality_fct_rejoining_of_upstream_concepts (seeds\dag\dag_seeds.yml)
23:56:14  Failure in test equality_fct_test_coverage (seeds\tests\tests_seeds.yml)
```

b-per commented:
It is not super easy to get the resulting duckdb database from the CI, and the logs won't give you much info.

But here is the diff for the 2 tests you mentioned when I run them locally

`fct_documentation_coverage`: 

| which\_diff | total\_models | documented\_models | documentation\_coverage\_pct | staging\_documentation\_coverage\_pct | intermediate\_documentation\_coverage\_pct | marts\_documentation\_coverage\_pct | other\_documentation\_coverage\_pct |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| a\_minus\_b | 19 | 4 | 21.05 | 20 | 0 | 0 | 75 |
| b\_minus\_a | 16 | 4 | 25 | 20 | 0 | 0 | 75 |

`fct_undocumented_models`:

| which\_diff | resource\_name | model\_type |
| :--- | :--- | :--- |
| a\_minus\_b | int\_model\_10 | intermediate |
| a\_minus\_b | int\_model\_11 | intermediate |
| a\_minus\_b | base\_model\_10 | base |

BradCr commented:
@b-per I updated my seeds `test_fct_docuimentation_coverage` and `test_fct_undocumented_models` to match the values you showed, then committed those changes. The tests on those models now pass on circleci (but fail on my local). Next it's failing in circleci on the `test_fct_chained_views_dependencies` test, even though that one passes on my local.

Since it fails fast and doesn't uncover all the errors, I'm worried that it's going to give me one or two errors every time it runs, I'll get your help on what the values are, fix the seed, and then it will give me the next errors and that this process could take a while, especially since I'm not getting the same failures locally.

Any thoughts on what I could do to speed that up? Did I do something to mess up my setup/installation, should I start over with a new branch and add my changes back in? I hate that I'm requiring so much hand-holding here, but I appreciate the help you've given me very much.

b-per commented:
We can help getting the test to pass, but, yes, it will be useful to understand why things are different locally for you 😄 

1. Could you do a `pip list` and share it here? I am especially interested in the versions of dbt-core, dbt-duckdb and duckdb itself. I am thinking that some of the changes might be related to some older DuckDB version
2. Could you share the config of your DuckDB connection in `profiles.yml`?
3. And/Or, could you find me on the dbt community Slack and share with me the DuckDB database file that you get after a `dbt run --full-refresh`? I can then try to figure out why things are different.

b-per commented on integration_tests/package-lock.yml:
Thinking out loud but I think that we'd want to remove this and add it to `.gitignore`

BradCr replied:
I noticed it created when I ran (I think) dbt deps. I will delete it and add it to `.gitignore`

b-per commented:
OK, so I spent a bit of time looking at the test results that need to be change but unfortunately, the new model under `integration_tests` requires changing the majority of the existing tests that we have.

For example, with the new model, there is now no violation of `fct_rejoining_of_upstream_concepts` anymore.

We should try to get your local version working but I am thinking that the test might need to be updated as well.

I am thinking that we might want to make the number of joins a variable, that defaults to 7, but that we could change to 3 in our `integration_tests` folder. Otherwise, it will impact way too many other models. (and I also think that people might want to configure themselves a number different from 7)

b-per commented:
Also, one of the seeds is the following:

```
resource_name,resource_type,file_path,join_count
fct_model_10,model,"models\\marts\\fct_model_10.sql",7
```

This is not going to work with CI as it runs on Linux and the file separator is the Windows one. We should just maybe remove checking that specific column.

This makes me think that some of the discrepancies you see when running locally might be due to you running on Windows when CI runs on Linux and we run those on MacOS at dbt Labs

BradCr commented:
Thank you for all the suggestions. Here is a list of what I've done and the info you asked for. I did the `dbt run --full-refresh` and then uploaded the duckdb database to my Google drive. I've messaged that link to you on Slack.

I changed the threshold for `fct_too_many_joins` to be a variable, defaulted to 7 in the main `dbt_project.yml`, and to 3 in the integration_tests.

I corrected the `test_fct_too_many_joins.csv` seed to use forward slashes (I had originally changed it to double-backslash because that's what my build put into the model in the duckdb.)

Contents of `pip list`, with dbt-code, dbt-duckdb, and duckdb at the top:
Package                   Version
------------------------- ------------
dbt-core                  1.7.5
dbt-duckdb                1.7.1
duckdb                    0.9.2
agate                     1.7.1
annotated-types           0.6.0
attrs                     23.2.0
Babel                     2.14.0
certifi                   2023.11.17
cffi                      1.16.0
charset-normalizer        3.3.2
click                     8.1.7
colorama                  0.4.6
dbt-extractor             0.5.1
dbt-semantic-interfaces   0.4.3
idna                      3.6
importlib-metadata        6.11.0
isodate                   0.6.1
Jinja2                    3.1.3
jsonschema                4.21.0
jsonschema-specifications 2023.12.1
leather                   0.3.4
Logbook                   1.5.3
MarkupSafe                2.1.3
mashumaro                 3.11
minimal-snowplow-tracker  0.0.2
more-itertools            10.2.0
msgpack                   1.0.7
networkx                  3.2.1
packaging                 23.2
parsedatetime             2.6
pathspec                  0.11.2
pip                       23.3.2
protobuf                  4.25.2
pycparser                 2.21
pydantic                  2.5.3
pydantic_core             2.14.6
python-dateutil           2.8.2
python-slugify            8.0.1
pytimeparse               1.1.8
pytz                      2023.3.post1
PyYAML                    6.0.1
referencing               0.32.1
requests                  2.31.0
rpds-py                   0.17.1
setuptools                65.5.0
six                       1.16.0
sqlparse                  0.4.4
text-unidecode            1.3
typing_extensions         4.9.0
urllib3                   1.26.18
zipp                      3.17.0

Duckdb `profiles.yml`:
```
integration_tests:
  target: duckdb
  outputs:
    duckdb:
      type: duckdb
      path: '/Users/bcron/python/duckdb/dbt_integration_tests.duckdb'
```

BradCr commented:
The fact that I haven't gotten the integration tests to pass is bugging me, so I decided to try something. I created a new branch in my repository to try and get a good, clean baseline test before making any changes. I installed the dbt-duckdb package, ran `dbt deps` in the root directory and in integration_tests, then ran `dbt build -f` in root and in `integration_tests`.

Without any modifications to any models or seeds, I am getting errors on 10 tests. I was able to resolve or explain all but one. Is it possible that some integration_tests seeds in `main` aren't up to date and need corrections in order to work? It looks like there have been other recent PRs that passed circleci tests just fine, so I've got to assume the issue is on my end, but then I don't understand why it would throw me errors on a new clone.

---

These three have the single forward slash vs. double backward slash issue, but models and paths are the same, so I imagine they would pass circleci just fine if I leave the seeds alone:
- `equality_fct_source_directories`
- `equality_fct_model_directories`
- `equality_fct_test_directories`

---

These tests had additional rows in the table that weren't in the seed. When I added those rows to the seeds, the tests pass just fine:

- `equality_fct_undocumented_models`
- `equality_fct_chained_views_dependencies`
- `equality_fct_model_fanout`
- `equality_fct_root_models`
- `equality_fct_missing_primary_key_tests`

---

This covergae test had different values in the seed vs. the table build by the model. When I changed the values in the seed, the test passes just fine:
- `equality_fct_documentation_coverage`

---

This covergae test had different values in the seed vs. the table built by the model. When I changed the values in the seed, there is still a weird error:
- `equality_fct_test_coverage`

After updating the `test_fct_test_coverage` seed, the `equality_fct_test_coverage` still fails. Here's where it gets interesting. The output of `equality_test_fct_test_coverage.sql` looks like this:

|which_diff|total_models|total_tests|tested_models|test_coverage_pct|test_to_model_ratio|staging_test_coverage_pct|intermediate_test_coverage_pct|marts_test_coverage_pct|other_test_coverage_pct|
|----------|------------|-----------|-------------|-----------------|-------------------|-------------------------|------------------------------|-----------------------|-----------------------|
|a_minus_b|18|13|6|**33.33000183105469**|0.7222|80.0|25.0|0.0|25.0|
|b_minus_a|18|13|6|**33.33**|0.7222|80.0|25.0|0.0|25.0|

Notice the `test_coverage_pct` from `a_minus_b` is showing in this query with 15 decimal places, rather than 2. But when I `select * from ` each table `test_fct_test_coverage` and `fct_test_coverage` independently, both of them show **33.33** as the value. Not sure what's happening here, but perhaps the circleci test would pass?

graciegoheen commented:
We saw the same rounding issue here https://github.com/dbt-labs/dbt-project-evaluator/pull/427#issuecomment-1941555822

BradCr commented:
> We saw the same rounding issue here [#427 (comment)](https://github.com/dbt-labs/dbt-project-evaluator/pull/427#issuecomment-1941555822)

Good to know it wasn't just me.

BradCr commented:
I saw there were a couple of PRs merged and my branch was out of date. I synched my branch, and it now fails the CircleCI tests. It was passing previously.

graciegoheen commented on models/marts/dag/fct_too_many_joins.sql:
```diff
@@ -0,0 +1,26 @@
+    from all_dag_relationships
+    where distance = 1
+    and child_model_type in ('marts')
+    group by
+        child,
+        child_resource_type,
+        child_file_path
```
```suggestion
    group by 1, 2
```

graciegoheen commented on models/marts/dag/fct_too_many_joins.sql:
```diff
@@ -0,0 +1,26 @@
+        count(distinct parent) as join_count
+    from all_dag_relationships
+    where distance = 1
+    and child_model_type in ('marts')
```
```suggestion
```
I think we should include all model types here

graciegoheen commented on models/marts/dag/fct_too_many_joins.sql:
```diff
@@ -0,0 +1,26 @@
+    select
+        *
+    from {{ ref('int_all_dag_relationships') }}
+    where not child_is_excluded
```
```suggestion
    where not child_is_excluded
    and child_resource_type = 'model'
```
I think we only want to include models here?

graciegoheen commented on models/marts/dag/fct_too_many_joins.sql:
```diff
@@ -0,0 +1,26 @@
+final as (
+    select
+        child as resource_name,
+        child_resource_type as resource_type,
```
```suggestion
```

graciegoheen commented on integration_tests/seeds/dag/test_fct_too_many_joins.csv:
```diff
@@ -0,0 +1,2 @@
+resource_name,resource_type,file_path,join_count
+fct_model_10,model,"models/marts/fct_model_10.sql",3
```
```suggestion
fct_model_10,"models/marts/fct_model_10.sql",3
```

graciegoheen commented on integration_tests/seeds/dag/test_fct_too_many_joins.csv:
```diff
@@ -0,0 +1,2 @@
+resource_name,resource_type,file_path,join_count
```
```suggestion
resource_name,file_path,join_count
```

graciegoheen commented on docs/rules/modeling.md:
```diff
@@ -429,3 +429,27 @@ or any other nested information.
+
+**How to Remediate**
+
+Bringing together a reasonable number (typically 4 to 6) of entities or concepts (staging models, or perhaps other intermediate models) that will be joined with another similarly purposed intermediate model to generate a mart. Rather than having too many joins in the mart, we can join two intermediate models that each house a piece of the complexity, giving us increased readability, flexibility,  ...
```
```suggestion
Bringing together a reasonable number (typically 4 to 6) of entities or concepts (staging models, or perhaps other intermediate models) that will be joined with another similarly purposed intermediate model to generate a mart. Rather than having too many joins, we can join two intermediate models that each house a piece of the complexity, giving us increased readability, flexibility, testing surface area, and insight into our components.
```

graciegoheen commented on integration_tests/seeds/dag/dag_seeds.yml:
```diff
@@ -90,4 +90,15 @@ seeds:
+          compare_model: ref('fct_too_many_joins')
+          compare_columns:
+            - resource_name
+            - resource_type
```
```suggestion
```

dave-connors-3 commented:
@BradCr thanks so much for the contribution! sorry this took so long!

BradCr commented:
@dave-connors-3 No problem, Dave. This was such a fun experience. I'm looking forward to contributing again.

Thank you also @b-per and @graciegoheen for all the help and guidance!
